### PR TITLE
3.x Add configuration to clustermgtd.conf for collecting compute console output.

### PR DESCRIPTION
**Pull request by davprat**
### Description of changes
* Generate configuration for compute console output logging in clustermgtd.conf on cluster creation
* Update configuration for compute console output logging in clustermgtd.conf on cluster update
* Configure CloudWatch Agent to publish compute console output logs from head node.
### Testing
* Manually tested happy path (Monitoring enabled) cluster creation.
* Manually tested CloudWatch Agent configuration by making sure new log file was correctly exported to CloudWatch.
* No testing on update yet. 
* No testing on Monitoring disabled.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by davprat** [comment]
Check failures due to know issue with stunnel package:
```
================================================================================
#20 567.0     Error executing action `create` on resource 'remote_file[/opt/parallelcluster/sources/stunnel-5.67.tar.gz]'
#20 567.0     ================================================================================
#20 567.0     
#20 567.0     OpenSSL::SSL::SSLError
#20 567.0     ----------------------
#20 567.0     SSL Error connecting to https://www.stunnel.org/downloads/stunnel-5.67.tar.gz - SSL_connect returned=1 errno=0 state=error: certificate verify failed (certificate has expired)
#20 567.0     
#20 567.0     Resource Declaration:
#20 567.0     ---------------------
#20 567.0     # In /etc/chef/local-mode-cache/cache/cookbooks/aws-parallelcluster-install/recipes/efs.rb
#20 567.0     
#20 567.0      91:   remote_file stunnel_tarball do
#20 567.0      92:     source node['cluster']['stunnel']['url']
#20 567.0      93:     mode '0644'
#20 567.0      94:     retries 3
#20 567.0      95:     retry_delay 5
#20 567.0      96:     not_if { ::File.exist?(stunnel_tarball) }
#20 567.0      97:   end
#20 567.0      98: 
```

File: attributes/default.rb
Comment (chenwany) [inline attributes/default.rb]: question: Could you explain the benefit of using multiple nested dict here to store the default value and enable? What about simply use `default['cluster']["compute_console_logging_max_sample_size"]` and `default['cluster']["compute_console_logging_enabled"]` because the clustermgtd configuration is also not nested?

File: cookbooks/aws-parallelcluster-config/files/default/cloudwatch_agent/cloudwatch_log_files.json
Comment (chenwany) [inline cookbooks/aws-parallelcluster-config/files/default/cloudwatch_agent/cloudwatch_log_files.json:160]: Is this `timestamp_format_key` manually tested? Asking because the console_output.logs are from system messge. And system message are in the format of
```
"timestamp_format_key": "month_first",
```

File: cookbooks/aws-parallelcluster-slurm/recipes/update_head_node.rb
Comment (chenwany) [inline cookbooks/aws-parallelcluster-slurm/recipes/update_head_node.rb]: Is the pattern need to be update here?

**Review by chenwany — COMMENTED** [review]
General question that not related to this PR: If there's a long lasting cluster, the logs will keep growing, how are we going to handle that? Is there a way to delete the logs when they are uploaded to the log group? Maybe we can leave it to when we go to log rotation story

File: attributes/default.rb
Comment (chenwany) [reply attributes/default.rb]: Should it be enable only when https://docs.aws.amazon.com/parallelcluster/latest/ug/Monitoring-v3.html#yaml-Monitoring-Logs-CloudWatch-Enabled is enable?

Comment (davprat) [reply attributes/default.rb]: It will use Monitoring-Logs-CloudWatch-Enabled if it is present, otherwise it defaults to True ([This line](https://github.com/aws/aws-parallelcluster-cookbook/pull/1598/files#diff-694753d857e75e10dfab299ef7cdb79106b6a6fb2b6812f637b490c9cc16b5d9R179)).

File: cookbooks/aws-parallelcluster-slurm/recipes/update_head_node.rb
Comment (davprat) [reply cookbooks/aws-parallelcluster-slurm/recipes/update_head_node.rb]: oof, yes you are correct.

**Review by delongmeng — COMMENTED** [review]
Just a general question, what tests (maybe just some manual tests) have been done for the code change in this PR?

File: cookbooks/aws-parallelcluster-config/files/default/cloudwatch_agent/cloudwatch_log_files.json
Comment (davprat) [reply cookbooks/aws-parallelcluster-config/files/default/cloudwatch_agent/cloudwatch_log_files.json:160]: I believe it is the same as the time format as clustermgtd - this entry is picking up the log file written out by a python logger that I've defined alongside clustermgtd's root logger and I've specified the format to be the same as clustermgtd.
So CloudWatch Agent will be parsing the event as written by clustermgtd and not parsing the contents (which is the system output) I think?

File: attributes/default.rb
Comment (davprat) [reply attributes/default.rb]: As to the first question, to me it seems more readable and kind of groups things by purpose, but I'm not too opinionated on this.

**Comment by davprat** [comment]
> General question that not related to this PR: If there's a long lasting cluster, the logs will keep growing, how are we going to handle that? Is there a way to delete the logs when they are uploaded to the log group? Maybe we can leave it to when we go to log rotation story

Yes, I was thinking of maybe adding log rotation for this specific log  to this story as a first step/experiment. I might do this after the beta goes out.

**Comment by davprat** [comment]
> Just a general question, what tests (maybe just some manual tests) have been done for the code change in this PR?

Good call out, I should have documented this in the PR.

File: cookbooks/aws-parallelcluster-config/files/default/cloudwatch_agent/cloudwatch_log_files.json
Comment (chenwany) [reply cookbooks/aws-parallelcluster-config/files/default/cloudwatch_agent/cloudwatch_log_files.json:160]: ok sounds good, let's test it with the node package PR

File: attributes/default.rb
Comment (davprat) [reply attributes/default.rb]: removed ['default'] sub-tree.

File: cookbooks/aws-parallelcluster-slurm/recipes/update_head_node.rb
Comment (davprat) [reply cookbooks/aws-parallelcluster-slurm/recipes/update_head_node.rb]: fixed

**Review by chenwany — DISMISSED** [review]
LGTM! lets test the timestamp for the log stream before merging it

**Review by chenwany — DISMISSED** [review]

**Review by chenwany — APPROVED** [review]

**Pull request merged**
